package rec.command;

// A list of commands that the user can issue.
message Command {
  enum Type {
    NONE = 0;

    // Pre-defined by Juce.
    QUIT = 0x1001;
    DEL = 0x1002;
    CUT = 0x1003;
    COPY = 0x1004;
    PASTE = 0x1005;
    SELECT_ALL = 0x1006;
    DESELECT_ALL = 0x1007;

    ADD_LOOP_POINT = 1;
    AUDIO_PREFERENCES = 2;
    CLEAR_NAVIGATOR = 3;
    CLEAR_LOOPS = 4;
    CLEAR_SAVED_FILE_DATA = 5;
    CLOSE_FILE = 6;

    DIM_VOLUME_TOGGLE = 77;
    EJECT_CDS = 7;

    INVERT_LOOP_SELECTION = 19;

    KEYBOARD_MAPPINGS = 18;
    MIDI_MAPPINGS = 17;

    MUTE_VOLUME_TOGGLE = 78;

    NUDGE_BEGIN_LEFT = 11;
    NUDGE_BEGIN_RIGHT = 12;
    NUDGE_END_LEFT = 13;
    NUDGE_END_RIGHT = 14;

    NUDGE_VOLUME_DOWN = 59;
    NUDGE_VOLUME_UP = 72;

    OPEN = 15;
    RECENT_FILES = 16;
    RESET_GAIN_TO_UNITY = 79;

    TOGGLE_WHOLE_SONG_LOOP = 58;
    TOGGLE_START_STOP = 50;
    TOGGLE_STRETCH_ENABLE = 51;

    TREE_CLOSE = 52;
    TREE_DOWN = 53;
    TREE_LEFT = 54;
    TREE_OPEN = 55;
    TREE_RIGHT = 56;
    TREE_UP = 57;

    ZOOM_IN = 70;
    ZOOM_OUT = 71;

    // Banks of 100,000 starting at 1,000,000 (and going back to 995).
    SELECT = 1000000;
    SELECT_ONLY = 1100000;
    TOGGLE = 1200000;
    UNSELECT = 1300000;
    JUMP = 1400000;
    JUMP_SELECTED  = 1500000;

    BANK_START = 1000000;
    BANK_END = 1600000;
    BANK_SIZE = 100000;

    LAST_COMMAND = 99;  // 89 is free.
  };

  optional Type command = 1;
};